# HG changeset patch
# User Juan Pablo Carbajal <ajuanpi+dev@gmail.com>
# Date 1386338921 -3600
#      Fri Dec 06 15:08:41 2013 +0100
# Branch stable
# Node ID b06675ef40f2f8707780824aa5951dd079d42514
# Parent  0f32b34f6ec0eef2489bfc22c13dce71410e52e9
interp1.m: Fix handling multi-column y-input with jumps (bug #40825).

* interp1.m: Eliminate jumps from both x AND y before doing
linear interpolation.

diff --git a/scripts/general/interp1.m b/scripts/general/interp1.m
--- a/scripts/general/interp1.m
+++ b/scripts/general/interp1.m
@@ -232,28 +232,33 @@ function yi = interp1 (x, y, varargin)
                  shiftdim (y, 1), szy(2:end));
       pp.orient = "first";
       if (ispp)
         yi = pp;
       else
         yi = ppval (pp, reshape (xi, szx));
       endif
     case "linear"
-      dy = diff (y);
-      dx = diff (x);
-      dx = repmat (dx, [1 size(dy)(2:end)]);
-      coefs = [(dy./dx).'(:), y(1:nx-1, :).'(:)];
+
       xx = x;
-
+      yy = y;
+      nxx = nx;
       if (have_jumps)
         ## Omit zero-size intervals.
-        coefs(jumps, :) = [];
+        yy(jumps, :) = [];
         xx(jumps) = [];
+        nxx = rows (xx);
       endif
 
+      dy = diff (yy);
+      dx = diff (xx);
+      dx = repmat (dx, [1 size(dy)(2:end)]);
+
+      coefs = [(dy./dx).'(:), yy(1:nxx-1, :).'(:)];
+
       pp = mkpp (xx, coefs, szy(2:end));
       pp.orient = "first";
 
       if (ispp)
         yi = pp;
       else
         yi = ppval (pp, reshape (xi, szx));
       endif
